certora/specs/dll-simple.spec

```diff
@@ -0,0 +1,292 @@
+methods {
+    getValueOf(address) returns (uint256) envfree
+    getHead() returns (address) envfree
+    getTail() returns (address) envfree
+    getNext(address) returns (address) envfree
+    getPrev(address) returns (address) envfree
+    remove(address) envfree
+    insertSorted(address, uint256) envfree
+    // added through harness
+    getInsertBefore() returns (address) envfree
+    getInsertAfter() returns (address) envfree
+    prevFromHead(address) returns (address) envfree
+    isForwardLinkedBetween(address, address) returns (bool) envfree
+    isDecrSortedFrom(address) returns (bool) envfree
+}
+
+// DEFINITIONS
+
+definition inDLL(address _id) returns bool =
+    getValueOf(_id) != 0;
+
+definition linked(address _id) returns bool =
+    getPrev(_id) != 0 || getNext(_id) != 0;
+
+definition isEmpty(address _id) returns bool =
+    ! inDLL(_id) && ! linked(_id);
+
+definition isTwoWayLinked(address first, address second) returns bool =
+    first != 0 && second != 0 => (getNext(first) == second <=> getPrev(second) == first);
+
+definition isHeadWellFormed() returns bool =
+    getPrev(getHead()) == 0 && (getHead() != 0 => inDLL(getHead()));
+
+definition isTailWellFormed() returns bool =
+    getNext(getTail()) == 0 && (getTail() != 0 => inDLL(getTail()));
+
+// INVARIANTS & RULES
+
+invariant zeroEmpty()
+    isEmpty(0)
+    filtered { f -> f.selector != insertSorted(address, uint256).selector }
+
+rule zeroEmptyPreservedInsertSorted(address _id, uint256 _value) {
+    env e; address prev;
+    address addr;
+
+    require isEmpty(0);
+    requireInvariant twoWayLinked(prev, getNext(prev));
+    requireInvariant noNextIsTail(prev);
+
+    insertSorted(_id, _value);
+
+    require prev == getInsertAfter();
+
+    assert isEmpty(0);
+}
+
+invariant headWellFormed()
+    isHeadWellFormed()
+    { preserved remove(address rem) {
+        requireInvariant zeroEmpty();
+        requireInvariant twoWayLinked(getPrev(rem), rem);
+        requireInvariant twoWayLinked(rem, getNext(rem));
+        requireInvariant linkedIsInDLL(getNext(rem));
+      }
+    }
+
+invariant tailWellFormed()
```

Comment:
Maybe add a comment

Reply:
I don't remember this, a comment about what ?

Reply:
about what it is verifying: construction and remove